Model/src/main/java/org/apidb/apicommon/model/report/ai/expression/AiExpressionCache.java

```diff
@@ -100,22 +146,69 @@ public AiExpressionCache(WdkModel wdkModel) throws IOException {
     _cache = new OnDiskCache(cacheParentDir, DEFAULT_TIMEOUT_MILLIS, DEFAULT_POLL_FREQUENCY_MILLIS);
   }
 
-  /**
-   * Tries to read a gene summary from the cache without populating if absent.
-   *
-   * @param summaryInputs inputs for cache lookup
-   * @return response JSON (indicating cache hit or not with data or miss reason respectively)
-   */
   public JSONObject readSummary(GeneSummaryInputs summaryInputs) {
-    try {
-      return _cache.visitContent(summaryInputs.getGeneId(),
-          geneDir -> getValidSummary(geneDir, summaryInputs));
+
+    // collect status of each experiment
+    int numComplete = 0;
+    Map<String, String> experiments = new LinkedHashMap<>();
+    for (ExperimentInputs datasetInput : summaryInputs.getExperimentsWithData()) {
+      Status status = getEntryStatus(datasetInput.getCacheKey(), datasetInput.getDigest());
+      if (status == Status.PRESENT) {
+        numComplete++;
+      }
+      experiments.put(datasetInput.getDatasetId(), status.val());
     }
-    catch (LookupException e) {
-      return e.toJson();
+
+    if (numComplete == summaryInputs.getExperimentsWithData().size()) {
+      // all experiments complete; check summary using short wait time
+      Status status = getEntryStatus(summaryInputs.getGeneId(), summaryInputs.getDigest());
+      if (status == Status.PRESENT) {
+        // all done; read result
+        try {
+          JSONObject summary = _cache.visitContent(summaryInputs.getGeneId(), dir -> getValidStoredData(dir, summaryInputs.getDigest()));
+          return new JSONObject()
+              .put(RESULT_STATUS_PROP, Status.PRESENT.val())
+              .put(SUMMARY_RESULT_PROP, summary);
+        }
+        catch (Exception e) {
+          // would not expect this since we already checked and are waiting for lock
```

Comment:
Couldn't something theoretically happen to the cached data in-between the `getEntryStatus` and the `visitContent`?

I don't understand how we are "waiting for lock" but maybe I've missed something.

I also wonder if we should make a `GeneSummaryInputs.getCacheKey()` function (that trivially calls `getGeneId()` just for symmetry/readability?